unixfs/src/dir/walk.rs

```diff
@@ -0,0 +1,777 @@
+                let (bytes, metadata, step) = IdleFileVisit::default()
+                    .start_from_parsed(flat, cache)?;
+
+                if let Some(visit) = step {
+                    Ok(Walk::MultiBlockFile {
+                        metadata,
+                        visit
+                    })
+                } else {
+                    Ok(Walk::SingleBlockFile {
+                        bytes,
+                        metadata
+                    })
+                }
+            },
+            UnixFsType::Metadata => todo!("metadata?"),
+            UnixFsType::Symlink => todo!("?"),
+        }
+    }
+
+    fn walk_directory<'a, I>(mut links: I, current: InnerEntry) -> Result<Walk<'a>, Error>
+        where I: Iterator<Item = Result<(Cid, String, usize), InvalidCidInLink>> + 'a,
+    {
+        if let Some(next) = links.next() {
+            let next = Some(next?);
+            let pending = links.collect::<Result<Vec<_>, _>>()?;
+
+            Ok(Walk::Walker(Walker {
+                current,
+                next,
+                pending,
+            }))
+        } else {
+            Ok(Walk::EmptyDirectory {
+                metadata: current.into(),
+            })
+        }
+    }
+
+    pub fn as_entry<'a>(&'a self) -> Entry<'a> {
+        self.current.as_entry()
+    }
+
+    pub fn pending_links(&self) -> (&Cid, impl Iterator<Item = &Cid>) {
+        use InnerKind::*;
+        let cids = self.pending.iter().map(|(cid, ..)| cid).rev();
+        match self.current.kind {
+            File(_, Some(ref visit)) => {
+                let (first, rest) = visit.pending_links();
+                (first, Either::Left(rest.chain(cids)))
+            },
+            _ => {
+                let next = self.next.as_ref()
+                    .expect("validated in start and continue_walk we have the next");
+                (&next.0, Either::Right(cids))
+            }
+        }
+    }
+
+    pub fn continue_walk<'a>(mut self, bytes: &'a [u8], cache: &mut Option<Cache>) -> Result<ContinuedWalk<'a>, Error> {
+        use InnerKind::*;
+
+        match &mut self.current.kind {
+            File(_, visit @ Some(_)) => {
+                let (bytes, step) = visit.take()
+                    .unwrap()
+                    .continue_walk(bytes, cache)?;
+
+                let file_continues = step.is_some();
+                *visit = step;
+
+                let segment = FileSegment::later(bytes, !file_continues);
+
+                let state = if file_continues || self.next.is_some() {
+                    State::Unfinished(self)
+                } else {
+                    State::Last(self.current)
+                };
+
+                return Ok(ContinuedWalk::File(segment, Item::from(state)))
+            },
+            _ => {}
+        }
+
+        let flat = FlatUnixFs::try_from(bytes)?;
+        let metadata = FileMetadata::from(&flat.data);
+
+        match flat.data.Type {
+            UnixFsType::Directory => {
```

Comment:
this and the hamtshard arms cannot be refactored into a method like in the `start` case. At least not without making the `self.next.take().expect("continued without next")`.